src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp

```diff
@@ -332,21 +331,130 @@ void VTKPolyDataWriterInterface::writeElementFields( vtkSmartPointer< vtkCellDat
+    // dimensions or more.
+
+    // First, check if the field requires a special treatment. For now, array4d< real64 > (for instance, for phaseCompFraction) are
+    // treated separately to make sure that they can be easily visualized in Paraview.
+    int numArrayDims = 0;
+
     er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
     {
       WrapperBase const & wrapper = esr.getWrapperBase( field );
-      writeField( wrapper, data, esr.size(), count );
+      numArrayDims = wrapper.numArrayDims();
     } );
-    celldata->AddArray( data );
+
+    // Then, output the field to VTK
+    // Note: fields with 5 dimensions or more are not supported and are not output to file below
+    if( numArrayDims < 4 )
+    {
+      // If the field has three dimensions or less, it can be output in a generic fashion
+      writeElementField1D2D3D< SUBREGION >( celldata, er, field );
+    }
+    else if( numArrayDims == 4 )
+    {
+      writeElementField4D< SUBREGION >( celldata, er, field );
+    }
+    else
+    {
+      GEOSX_ERROR( "Fields with more than 4 dimensions cannot be output to VTK" );
+    }
   }
 }
+
+template< class SUBREGION >
+void VTKPolyDataWriterInterface::writeElementField1D2D3D( vtkSmartPointer< vtkCellData > const celldata,
+                                                          ElementRegionBase const & er,
+                                                          string const & field ) const
+{
+  localIndex count = 0;
+
+  vtkSmartPointer< VTKGEOSXData > data = VTKGEOSXData::New();
+  data->SetNumberOfValues( er.getNumberOfElements< SUBREGION >() );
+  data->SetName( field.c_str() );
+
+  er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
+  {
+    WrapperBase const & wrapper = esr.getWrapperBase( field );
+    writeField( wrapper, data, esr.size(), count );
+  } );
+  celldata->AddArray( data );
+}
+
+template< class SUBREGION >
+void VTKPolyDataWriterInterface::writeElementField4D( vtkSmartPointer< vtkCellData > const celldata,
+                                                      ElementRegionBase const & er,
+                                                      string const & field ) const
+{
+  // Step 1: we need to figure out the size of the two dimensions of the array that we are going to output component by component
+  localIndex nCompInSecondDim = 0;
+  localIndex nCompInThirdDim = 0;
+  er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
+  {
+    WrapperBase const & wrapper = esr.getWrapperBase( field );
+    Wrapper< array4d< real64 > > const & wrapperT = reinterpret_cast< Wrapper< array4d< real64 > > const & >( wrapper );
+    arrayView4d< real64 const > const & sourceArray = wrapperT.reference();
+    sourceArray.move( LvArray::MemorySpace::host, true );
+    arraySlice4d< real64 const > const & tmp = sourceArray.toSliceConst();
+    arraySlice4d< real64 const, -1 > const & sourceSlice = reinterpret_cast< arraySlice4d< real64 const, -1 > const & >( tmp );
```

Comment:
Would you mind putting this whole sequence in a function (just a top-level function in this cpp would be fine, perhaps in an unnamed namespace) and adding a warning comment that this is a temporary hack to get around the lack of type dispatching for 4D arrays?

Reply:
Hi @klevzoff I don't mind doing what you suggest but I got stuck with the C++ casting syntax for the "return by value":

I can do something like that:
```
arraySlice4d< real64 const, -1 >                                                                                                                                                         
get4DSliceFromWrapper( WrapperBase const & wrapper )                                                                                                                                     
{                                                                                                                                                                                        
  Wrapper< array4d< real64 > > const & wrapperT = reinterpret_cast< Wrapper< array4d< real64 > > const & >( wrapper );                                                                   
  arrayView4d< real64 const > const & sourceArray = wrapperT.reference();                                                                                                                
  sourceArray.move( LvArray::MemorySpace::host, false );                                                                                                                                 
  arraySlice4d< real64 const > const & tmp = sourceArray.toSliceConst();                                                                                                                 
  arraySlice4d< real64 const, -1 > const & sourceSlice = reinterpret_cast< arraySlice4d< real64 const, -1 > const & >( tmp );                                                                                                                                                                                         
  return sourceSlice;                                                                                                                                                                    
}
```
but it does not work (it crashes when I access the returned slice) because I assume my `sourceSlice` is a reference to a temporary (?). Ideally I would like to do something like that:
```
arraySlice4d< real64 const, -1 >                                                                                                                                                         
get4DSliceFromWrapper( WrapperBase const & wrapper )                                                                                                                                     
{                                                                                                                                                                                        
  Wrapper< array4d< real64 > > const & wrapperT = reinterpret_cast< Wrapper< array4d< real64 > > const & >( wrapper );                                                                   
  arrayView4d< real64 const > const & sourceArray = wrapperT.reference();                                                                                                                
  sourceArray.move( LvArray::MemorySpace::host, false );                                                                                                                                 
  arraySlice4d< real64 const > const tmp = sourceArray.toSliceConst();                                                                                                                 
  arraySlice4d< real64 const, -1 > const sourceSlice = reinterpret_cast< arraySlice4d< real64 const, -1 > const >( tmp );                                                                                                                                                                                         
  return sourceSlice;                                                                                                                                                                    
}
```
but this does not compile:
```
reinterpret_cast from 'const arraySlice4d<const geosx::real64>' (aka 'const ArraySlice<const double, 4, 3, long>') to 'const arraySlice4d<const geosx::real64, -1>' (aka 'const ArraySlice<const double, 4, -1, long>') is not allowed 
```
I have to say I am a little bit confused by the syntax in this piece of code. Is there a better way to safely return the slice by value?

Reply:
The first attempt seems like it should work. You can try removing reference from `tmp`, and also initializing a local variable of type `arraySlice4d< real64 const, -1 >` which you then return (instead of having the compiler copy-construct the return value for you). Try printing the slice object before returning to make sure it's initialized there.

If none of it works, we've likely been hit by the compiler optimizing around UB. Specifically, it likely just deletes all the code following an illegal `reinterpret_cast` and skips initializing the return value (it is allowed to do so by the standard). I see two ways of working around it:
1. Declare a function taking a `void *` and returning a `void *` and pass the slice through it, using two `reinterpret_cast`s for the argument and the return value. The function's definition must be in a separate translation unit. This approach keeps the UB, but prevents the compiler from seeing through it. I don't like this any better than I did the original solution, but it's an option.
2. (Preferred) Add a (conditionally enabled) explicit conversion constructor to `ArraySlice`, to allow constructing a slice with negative `USD` from a slice with any other `USD`. This is the route we should've probably taken from the start, but it's a bit longer.

Reply:
Thanks! I tried the first part of what you suggested (removing the reference and adding the local variable), I see that the slice is well initialized in the function doing the cast (`get4DSliceFromWrapper`), but the slice that I get after the function has returned is garbage. The code only works properly when the two `reinterpret_cast` are in the body of the main function that writes the field (`writeElementField4D`). 

It would be great to have option 2, but I am not sure I can write it on my own. I can try to backtrack and add the `array4d` + permutations to the supported datatypes. I started doing that once, it requires the addition of a few functions in `FieldSpecifications` as you said, but maybe it is going to be a better strategy.

Reply:
That sounds good to me. I can also help write the `ArraySlice` constructor, but extending the switchyard has always been the best option (as long as you don't encounter some existing call site that hard-assumed no more than 3 dims).

By the way, I looked again at the code above... I think the culprit was `sourceArray` is a local temporary, and any slice produced from it contained pointers to that temporary, which went out of scope. I thought initially it was a reference to the original (permanent) array object, but there is a non-`const` to `const` conversion involved which now produces a new temporary view. In any case, I'm actually relieved we won't be going the UB route in the end.

Reply:
I've tried to get rid of as many instances of this bug as possible (creating a slice from a temporary view) and I've managed to make it a compilation error for rvalues, but I don't think preventing this bug is possible. Improving documentation would help.